File: java/IsEqual.java
public class IsEqual{
  public static void main(String[] args){
    System.out.println(isEqual1(42,42));    // true
    System.out.println(isEqual2(42,42));    // true
    Integer x = new Integer(42);
    Integer y = new Integer(42);
    System.out.println(isEqual2(x,y));      // false 
  }

  private static boolean isEqual1(int x, int y){
      return x == y;
  }

  private static boolean isEqual2(Integer x, Integer y){
    return x == y;
  }
}

File: scala/IsEqual.scala
// equality in scala:
// == is usual equality for value types (subclassing AnyVal)
// == cannot be overriden and is an alias of equals which can be overriden
// for reference types (subclassing AnyRef). equals is by default implemented
// as reference identity.
// there is the method 'eq' which cannot be overriden and always refer to
// reference identity for reference types

def isEqual1(x: Int, y: Int) = x == y
def isEqual2(x: Any, y: Any) = x == y

println(isEqual1(42,42))
println(isEqual2(42,42))


